### DISCUSSION TITLE
Possible wrong in the formula RSI indicator?

### DISCUSSION THREAD

**Discussion by morycom** [DaveSkender/Stock.Indicators#215] [Q&A] [answered]
Hi Dave
I think the formula used in the rsi indicator is wrong.
The rsi calculated for a particular day changes when the number of input data increases or decreases.
For example, if we use the price history of the last 20 trading days to calculate, the rsi calculated for the first day is different from when we use the price history of the last 30 trading days.
However, according to the rsi formula, it does not depend on the number of input data.
thanks a lot .

**Comment by DaveSkender** [comment, marked as answer]
Since [Relative Strength Index](https://daveskender.github.io/Stock.Indicators/indicators/Rsi/#content) uses a smoothing technique, there is essentially a "guess" value at the start of the historical quotes series that converges in accuracy over a warmup period.  As a result, yes, if you're using 20 vs 30 historical quotes for RSI(20), for example, you'd get different RSI values for early periods in your series.  Because of this, we recommend you use 250 preceding historical quotes for the best accuracy.  See also the discussion on Indicator Series vs Increments (#216).

**Reply by DaveSkender** [reply]
You can test this theory by using incrementally more `history` to see how a specific day's value converges.  A history of 20-30 preceding periods will have more variance, but if you compare the same day using 250 vs 300 preceding periods of history, you'll only detect 0.0000x differences.

You'll have to determine for yourself how much variance is acceptable and provide more or less history accordingly.

**Reply by DaveSkender** [reply]
I'll likely increase the `BadHistoryException` threshold on RSI to prevent the use of fewer than 50 periods of `history`, for this reason.  The current threshold only requires you to provide `N` periods of history, where `N` is the lookback period. (#222)

I'm also trying to decide if we should simply exclude those first 100 values in the results as well.  The reason I've kept them is that some, and possibly most use cases do not consider 0.0x and 0.00x convergence errors a problem.

**Reply by DaveSkender** [reply]
I've been running some tests to try to determine how fast some of these indicators converge.  It can be quite bad if you do not provide enough history where convergence is a factor.  Here's an example of RSI(14):

```txt
RSI(14) on 12/31/2018 with  20 periods of history: 34.90454144
RSI(14) on 12/31/2018 with  50 periods of history: 42.09491541
RSI(14) on 12/31/2018 with  75 periods of history: 42.06831775
RSI(14) on 12/31/2018 with 100 periods of history: 42.07401334
RSI(14) on 12/31/2018 with 120 periods of history: 42.07779927
RSI(14) on 12/31/2018 with 150 periods of history: 42.07757013
RSI(14) on 12/31/2018 with 200 periods of history: 42.07731651
RSI(14) on 12/31/2018 with 250 periods of history: 42.07732365
RSI(14) on 12/31/2018 with 350 periods of history: 42.07732357
RSI(14) on 12/31/2018 with 500 periods of history: 42.07732357
```
Does anybody have any opinions on how the library should handle it?  Basically, there are three options that I can see:

1. **Do nothing.**  Provide documented guidance, but let users decide how much to provide.  This option leaves open some potentially poor quality data for users who are not aware of this convergence and supply only the bare minimum.  _This is the current state._
2. **Increase BadHistoryException threshold**, but still provide the full results dataset, including the first X periods of yet-to-converge data.
3. **Exclude the first X periods** from the results, based on analysis of a certain convergence threshold.

**Reply by DaveSkender** [reply]
After doing some more analysis, it looks like the sweet spot for the optimal amount of `history` to provide for RSI is `10×N` periods before the intended usage period, where `N` is the specified lookback period.

**Reply by DaveSkender** [reply]
This is making me think that the standard simple moving average initial guess value has room for improvement.  When the initial guess is farther off from the ideal RSI value, it takes longer to converge.  This seems to be what happens in longer lookback periods.  With that said, I don't think RSI makes much sense with a lookback period greater than 25 or so; and if you want to do long time horizons, move up to weekly or monthly data instead of using larger lookback periods.

**Reply by DaveSkender** [reply]
Some history on RSI here, for the data nerds like me out there:
https://en.wikipedia.org/wiki/Relative_strength_index

There's a section on Cutler's RSI, which is different than the standard Wilder's RSI that we use.  Cutler was trying to solve this problem of convergence vs history length.
